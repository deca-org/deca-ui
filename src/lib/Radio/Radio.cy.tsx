import { standardColors } from '../Theme';
import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Radio from './Radio';

describe('components/Radio', () => {
  describe('before click', () => {
    it('border-color', () => {
      cy.mount(<Radio label="Label" />);

      cy.get('label')
        .before('border-color')
        .should('eq', Test.color('gray600'));
    });
    it('disabled', () => {
      cy.mount(<Radio label="Label" disabled />);

      cy.get('label').should('have.css', 'color', Test.color('gray500'));

      cy.get('label')
        .before('border-color')
        .should('eq', Test.color('gray400'));
    });
  });
  describe('colors', () => {
    it('label color', () => {
      cy.mount(<Radio label="Label" />);
      cy.get('label').should('have.css', 'color', Test.color('black'));
    });
    standardColors.map((color) => {
      describe(color, () => {
        it('background-color', () => {
          cy.mount(<Radio color={color} label="Label" />);

          cy.get('label').click();

          // wait for css transition to finish
          cy.wait(250);
          cy.get('.decaRadio-circle').should(
            'have.css',
            'background-color',
            Test.color(color)
          );
        });
        it('border-color', () => {
          cy.mount(<Radio color={color} label="Label" />);

          cy.get('label').click();

          // wait for css transition to finish
          cy.wait(250);
          cy.get('label')
            .before('border-color')
            .should('eq', Test.color(color));
        });
        it('disabled', () => {
          cy.mount(
            <Radio color={color} label="Label" initialSelect disabled />
          );

          cy.get('label').before('opacity').should('eq', '0.6');
          cy.get('.decaRadio-circle').should('have.css', 'opacity', '0.5');
        });
      });
    });
  });
  describe('sizes', () => {
    it('sm', () => {
      cy.mount(<Radio size="sm" label="Label" />);
      cy.get('label').before('width').should('eq', Test.size('2'));
      cy.get('label').before('height').should('eq', Test.size('2'));
      cy.get('label').before('margin-right').should('eq', Test.space('1'));
      cy.get('label').should('have.css', 'font-size', Test.fontSize('caption'));
      cy.get('.decaRadio-circle').should('have.css', 'width', Test.size('1'));
    });
    it('md', () => {
      cy.mount(<Radio size="md" label="Label" />);
      cy.get('label').before('width').should('eq', Test.size('3'));
      cy.get('label').before('height').should('eq', Test.size('3'));
      cy.get('label').before('margin-right').should('eq', Test.space('2'));
      cy.get('label').should('have.css', 'font-size', Test.fontSize('bodySm'));
      cy.get('.decaRadio-circle').should('have.css', 'width', Test.size('2'));
    });
    it('lg', () => {
      cy.mount(<Radio size="lg" label="Label" />);
      cy.get('label').before('width').should('eq', Test.size('4'));
      cy.get('label').before('height').should('eq', Test.size('4'));
      cy.get('label').before('margin-right').should('eq', Test.space('2'));
      cy.get('label').should('have.css', 'font-size', Test.fontSize('body'));
      cy.get('.decaRadio-circle').should('have.css', 'width', Test.size('3'));
    });
  });

  describe('no label', () => {
    describe('should have no margin', () => {
      it('sm', () => {
        cy.mount(<Radio size="sm" />);
        cy.get('label').before('margin-right').should('eq', '0px');
      });
      it('md', () => {
        cy.mount(<Radio size="md" />);
        cy.get('label').before('margin-right').should('eq', '0px');
      });
      it('sm', () => {
        cy.mount(<Radio size="md" />);
        cy.get('label').before('margin-right').should('eq', '0px');
      });
    });
  });

  describe('dark mode', () => {
    it('label color', () => {
      cy.darkMount(<Radio label="Label" />);
      cy.get('label').should('have.css', 'color', Test.color('white'));
    });

    it('disabled state', () => {
      cy.darkMount(<Radio label="Label" initialSelect disabled />);
      cy.get('label').before('opacity').should('eq', '0.35');
      cy.get('.decaRadio-circle').should('have.css', 'opacity', '0.35');
    });
  });
});
