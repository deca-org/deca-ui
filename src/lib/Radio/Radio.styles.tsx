import { styled } from '@lib/Theme/stitches.config';
import { getStaticColor } from '@lib/Utils';
import { darken } from 'polished';

export const StyledRadioWrapper = styled('div', {
  position: 'relative',
  boxSizing: 'border-box',
  display: 'flex',
  alignItems: 'center',
  fontFamily: '$normal',
  transition: '$default',
});

export const StyledRadio = styled('input', {
  cursor: 'pointer',
  opacity: 0,
  position: 'absolute',
  transition: '$default',
  '&:checked + label > div': {
    visibility: 'visible',
  },
  compoundVariants: [
    {
      isDisabled: true,
      isDark: false,
      css: {
        pointerEvents: 'none',
        '& + label::before': {
          opacity: '60%',
        },
        '&:checked + label > div': {
          opacity: '50%',
        },
      },
    },
    {
      isDisabled: true,
      isDark: true,
      css: {
        pointerEvents: 'none',
        '& + label::before': {
          opacity: '35%',
        },
        '&:checked + label > div': {
          opacity: '35%',
        },
      },
    },
  ],
  variants: {
    size: {
      sm: {},
      md: {},
      lg: {},
    },
    color: {
      primary: {
        '&:checked + label > div': {
          bg: '$primary',
        },
        '&:checked + label::before': {
          borderColor: '$primary',
        },
        '&:checked:active + label::before': {
          borderColor: '$primary-darken-1',
        },
      },
      secondary: {
        '&:checked + label > div': {
          bg: '$secondary',
        },
        '&:checked + label::before': {
          borderColor: '$secondary',
        },
        '&:checked:active + label::before': {
          borderColor: '$secondary-darken-1',
        },
      },
      success: {
        '&:checked + label > div': {
          bg: '$success',
        },
        '&:checked + label::before': {
          borderColor: '$success',
        },
        '&:checked:active + label::before': {
          borderColor: '$success-darken-1',
        },
      },
      warning: {
        '&:checked + label > div': {
          bg: '$warning',
        },
        '&:checked + label::before': {
          borderColor: '$warning',
        },
        '&:checked:active + label::before': {
          borderColor: '$warning-darken-1',
        },
      },
      error: {
        '&:checked + label > div': {
          bg: '$error',
        },
        '&:checked + label::before': {
          borderColor: '$error',
        },
        '&:checked:active + label::before': {
          borderColor: '$error-darken-1',
        },
      },
    },

    isDisabled: {
      true: {},
      false: {
        '&:focus-visible + label::before': {
          boxShadow: '$a11y',
        },
      },
    },
    isDark: {
      true: {},
      false: {},
    },
  },
});

export const StyledRadioLabel = styled('label', {
  color: '$text',
  position: 'relative',
  cursor: 'pointer',
  display: 'flex',
  alignItems: 'center',
  transition: '$default',
  '&::before': {
    boxSizing: 'content-box',
    borderColor: '$gray600',
    br: '$pill',
    borderStyle: 'solid',
    borderWidth: '$normal',
    transition: '$default',
    content: '',
  },
  '& div': {
    br: '$pill',
    position: 'absolute',
    transition: '$default',
    visibility: 'hidden',
  },
  compoundVariants: [
    {
      isDisabled: true,
      isDark: true,
      css: {
        color: '$gray700',
      },
    },
    {
      isDisabled: true,
      isDark: false,
      css: {
        color: '$gray500',
      },
    },
    {
      hasLabel: true,
      size: 'sm',
      css: {
        '&::before': {
          mr: '$1',
        },
      },
    },
    {
      hasLabel: true,
      size: 'md',
      css: {
        '&::before': {
          mr: '$2',
        },
      },
    },
    {
      hasLabel: true,
      size: 'lg',
      css: {
        '&::before': {
          mr: '$2',
        },
      },
    },
  ],
  variants: {
    size: {
      sm: {
        fontSize: '$caption',
        '& div': {
          size: '$1',
          ml: '$1',
          mt: 'calc($0 / 4)',
        },
        '&::before': {
          size: '$2',
          mt: 'calc($0 / 4)',
        },
      },
      md: {
        fontSize: '$bodySm',
        '& div': {
          size: '$2',
          ml: '$1',
        },
        '&::before': {
          size: '$3',
        },
      },
      lg: {
        fontSize: '$body',
        '& div': {
          size: '$3',
          ml: '$1',
        },
        '&::before': {
          size: '$4',
        },
      },
    },
    color: {
      primary: {},
      secondary: {},
      success: {},
      warning: {},
      error: {},
    },
    isDisabled: {
      true: {
        cursor: 'not-allowed',
        color: '$gray500',
        '-webkit-user-select': 'none',
        '-moz-user-select': 'none',
        '-ms-user-select': 'none',
        userSelect: 'none',
        '&::before': {
          borderColor: '$gray400',
        },
      },
      false: {
        '&:hover::before': {
          transition: '$default',
          borderColor: darken(0.125, getStaticColor('gray600')),
        },
      },
    },
    hasLabel: {
      true: {},
      false: {},
    },
    isDark: {
      true: {},
      false: {},
    },
  },
});
