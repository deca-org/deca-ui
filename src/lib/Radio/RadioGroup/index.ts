import RadioGroup from './RadioGroup';
export * from './RadioGroup';
export { RadioGroup };
export default RadioGroup;
