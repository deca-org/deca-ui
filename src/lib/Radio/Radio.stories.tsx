import { DecaUIProvider } from '@lib/Theme';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Radio from './Radio';

export default {
  title: 'Radio',
  component: Radio,
} as ComponentMeta<typeof Radio>;

const Template: ComponentStory<typeof Radio> = (args) => <Radio {...args} />;

export const Default = Template.bind({});
Default.args = {
  label: 'Label',
  size: 'md',
  color: 'primary',
  disabled: false,
  css: {},
  className: '',
};

export const NoLabel = Template.bind({});
NoLabel.args = {
  ...Default.args,
  label: '',
};

export const WithTheme = Template.bind({});

WithTheme.args = { ...Default.args };
WithTheme.decorators = [
  (Story) => (
    <DecaUIProvider
      theme={{
        colors: {
          primary: '$green600',
        },
      }}
    >
      <Story />
    </DecaUIProvider>
  ),
];

export const DarkMode = Template.bind({});

DarkMode.args = { ...Default.args };
DarkMode.parameters = { backgrounds: { default: 'dark' } };

DarkMode.decorators = [
  (Story) => (
    <DecaUIProvider mode="dark">
      <Story />
    </DecaUIProvider>
  ),
];
