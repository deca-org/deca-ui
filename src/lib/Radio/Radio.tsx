import { ThemeContext } from '@lib/Theme';
import { CSS, StandardColors } from '@lib/Theme/stitches.config';
import {
  Modify,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
  uuid,
  __DEV__,
  MasterComponent,
} from '@lib/Utils';
import clsx from 'clsx';
import React, { useMemo } from 'react';

import {
  StyledRadioWrapper,
  StyledRadio,
  StyledRadioLabel,
} from './Radio.styles';

import RadioGroup from './RadioGroup';

/**
 * Radio buttons allow the user to select one option from a set.
 */
interface Props
  extends Modify<
    React.ComponentPropsWithRef<'input'>,
    {
      /**
       * Size of the component.
       * @default md
       */
      size?: 'sm' | 'md' | 'lg';
    }
  > {
  /**
   * Size of the component.
   * @default md
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * Text label for the component.
   */
  label?: string;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Disabled state applied to the component.
   * @default false
   */
  disabled?: boolean;
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * Color of radio button when active.
   * @default primary
   */
  color?: StandardColors;
  /**
   * The value assigned to the radio button.
   */
  value?: string;
  /**
   * Callback fired when a radio button is selected
   */
  onChange?(e: React.ChangeEvent<HTMLInputElement>): void;
  /**
   * Callback fired when a radio button is focused
   */
  onFocus?(e: React.ChangeEvent<HTMLInputElement>): void;
  /**
   * Name is used as an identifier in a form.
   */
  name?: string;
  /**
   *  Whether or not the checkbox is initially checked.
   */
  initialSelect?: boolean;
  /**
   * Whether or not the radio button is selected.
   */
  selected?: boolean;
}

export type RadioProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type RadioComponent = (<C extends React.ElementType = 'input'>(
  props: RadioProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Radio: RadioComponent = React.forwardRef(
  <T extends React.ElementType = 'input'>(
    {
      size = 'md',
      label,
      className,
      disabled = false,
      as,
      css,
      color = 'primary',
      value,
      name,
      onChange,
      onFocus,
      initialSelect,
      selected,
      ...props
    }: RadioProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const radioId = useMemo(() => {
      if (props.id) {
        return props.id;
      } else {
        return uuid();
      }
    }, [props.id]);

    const preClass = 'decaRadio';

    const changeHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (disabled) return;
      onChange && onChange(e);
    };

    const focusHandler = (e: React.ChangeEvent<HTMLInputElement>) => {
      if (disabled) return;
      onFocus && onFocus(e);
    };

    const { dark } = React.useContext(ThemeContext);

    return (
      <StyledRadioWrapper className={clsx(className, `${preClass}-root`)}>
        <StyledRadio
          id={radioId}
          type="radio"
          ref={ref}
          className={`${preClass}-input`}
          size={size}
          color={color}
          disabled={disabled}
          isDisabled={disabled}
          value={value}
          name={name}
          onChange={changeHandler}
          onFocus={focusHandler}
          isDark={dark}
          {...(initialSelect && { defaultChecked: initialSelect })}
          {...(!initialSelect && { checked: selected })}
          {...props}
        />
        <StyledRadioLabel
          htmlFor={radioId}
          css={css}
          size={size}
          as={as as any} // as prop will not accept T in this case
          color={color}
          isDisabled={disabled}
          hasLabel={label ? true : false}
          isDark={dark}
        >
          <div className={`${preClass}-circle`} />
          {label && label}
        </StyledRadioLabel>
      </StyledRadioWrapper>
    );
  }
);

if (__DEV__) {
  Radio.displayName = 'DecaUI.Radio';
}

export default Radio as MasterComponent<
  HTMLInputElement,
  RadioProps<React.ElementType>,
  {
    Group: typeof RadioGroup;
  }
>;
