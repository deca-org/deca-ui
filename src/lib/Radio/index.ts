import Radio from './Radio';
import RadioGroup from './RadioGroup';

Radio.Group = RadioGroup;

export * from './Radio';
export { Radio };
export default Radio;
