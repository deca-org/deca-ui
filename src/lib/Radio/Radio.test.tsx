import { render } from '@testing-library/react';
import React from 'react';

import Radio from './Radio';

describe('components/Radio', () => {
  it('matches snapshot', () => {
    const { asFragment } = render(<Radio label="Label Text" />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders all colors', () => {
    const { asFragment } = render(
      <>
        <Radio color="primary" />
        <Radio color="secondary" />
        <Radio color="success" />
        <Radio color="warning" />
        <Radio color="error" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders all sizes', () => {
    const { asFragment } = render(
      <>
        <Radio size="sm" />
        <Radio size="md" />
        <Radio size="lg" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
