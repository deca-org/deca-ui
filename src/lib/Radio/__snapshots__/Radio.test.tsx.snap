// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Radio matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-dpPCZp-color-primary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false c-jponNy-jQVVMl-cv"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
      Label Text
    </label>
  </div>
</DocumentFragment>
`;

exports[`components/Radio renders all colors 1`] = `
<DocumentFragment>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-dpPCZp-color-primary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-gxpMwf-color-secondary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-bYSuiK-color-success c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-cPFsdX-color-warning c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-cKfDIh-color-error c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
</DocumentFragment>
`;

exports[`components/Radio renders all sizes 1`] = `
<DocumentFragment>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-dpPCZp-color-primary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-biyOBk-size-sm c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-dpPCZp-color-primary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-cCGZrG-size-md c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
  <div
    class="c-cDEphG decaRadio-root"
  >
    <input
      class="c-iIDhYW c-iIDhYW-dpPCZp-color-primary c-iIDhYW-kueLmT-isDisabled-false decaRadio-input"
      id="1"
      type="radio"
      value=""
    />
    <label
      class="c-jponNy c-jponNy-fNoFLO-size-lg c-jponNy-hqEieG-isDisabled-false"
      for="1"
    >
      <div
        class="decaRadio-circle"
      />
    </label>
  </div>
</DocumentFragment>
`;
